lib/Sema/CSBindings.cpp

```diff
@@ -632,6 +632,19 @@ ConstraintSystem::getPotentialBindings(TypeVariableType *typeVar) {
         result.FullyBound = true;
       }
       break;
+
+    case ConstraintKind::OneWayBind:
+    case ConstraintKind::OneWayBindParam: {
```

Comment:
My 2 cents - I think `BindParam` is really the same as `OneWayBindParam` and doesn't require special constraint, since we already have special logic in `getPotentialBindings` to avoid producing results for type variable on its right-hand side until left is bound or there is no other way (currently it just sets `PotentiallyIncomplete` flag with de-prioritizes it), if the `EnableSolverOneWayConstraints` flag is set `BindParam` could be treated specially in constraint graph as well.

 For `OneWayBind` since it's applicable to closure result type it would be great (for diagnostics in particular) to actually have it called `ClosureResult`, similar to `FunctionResult` constraint. I think having a `ClosureResult` constraint could be beneficial in general even if not for uni-directional type-checking, because it makes it easier to identify contextual type mismatches related to closure result type.

Reply:
Now that I've added `Builtin.one_way`, I think I'll remove `OneWayBindParam` and the command-line flag completely. A wholesale change to the way type inference works with closures was an interesting experiment, but it doesn't belong in the tree. We can roll out and test this functionality more narrowly.

Reply:
Note that `OneWayBindParam` is now gone.